pkg/webhook/defaults/pods.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2020-2021 Clastix Labs
+// SPDX-License-Identifier: Apache-2.0
+
+package defaults
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	corev1 "k8s.io/api/core/v1"
+	schedulev1 "k8s.io/api/scheduling/v1"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	capsulev1beta2 "github.com/clastix/capsule/api/v1beta2"
+	"github.com/clastix/capsule/pkg/webhook/utils"
+)
+
+func mutatePodDefaults(ctx context.Context, req admission.Request, c client.Client, decoder *admission.Decoder, recorder record.EventRecorder) *admission.Response {
+	var err error
+
+	pod := &corev1.Pod{}
+	if err = decoder.Decode(req, pod); err != nil {
+		response := admission.Errored(http.StatusInternalServerError, err)
+
+		return &response
+	}
+
+	var tnt *capsulev1beta2.Tenant
+
+	tnt, err = utils.TenantByStatusNamespace(ctx, c, pod.Namespace)
+	if err != nil {
+		return utils.ErroredResponse(err)
+	}
+
+	if tnt == nil {
+		return nil
+	}
+
+	allowed := tnt.Spec.PriorityClasses
+
+	if allowed == nil || allowed.Default == "" {
+		return nil
+	}
+
+	priorityClassPod := pod.Spec.PriorityClassName
+	mutate := false
+
+	var cpc *schedulev1.PriorityClass
+
+	// PriorityClass name is empty, if no GlobalDefault is set and no PriorityClass was given on pod
+	if priorityClassPod != "" {
+		if priorityClassPod != allowed.Default {
+			cpc, err = utils.GetPriorityClassByName(ctx, c, priorityClassPod)
+			// Should not happen, since API already checks if PC present
+			if err != nil {
+				response := admission.Denied(NewPriorityClassError(priorityClassPod, err).Error())
+
+				return &response
+			}
+		}
+	} else {
+		mutate = true
+	}
+
+	// If given class is cluster default
+	if cpc != nil && utils.IsDefaultPriorityClass(cpc) {
+		// If cluster default is not allowed in tenant
+		if !tnt.Spec.PriorityClasses.MatchSelectByName(cpc) {
+			mutate = true
+		}
+	}
+
+	if mutate {
+		var pc *schedulev1.PriorityClass
+
+		pc, err = utils.GetPriorityClassByName(ctx, c, allowed.Default)
+		if err != nil {
+			msg := fmt.Sprintf("Failed to assign tenant default Priority Class: %s", err)
+			response := admission.Denied(msg)
+
+			return &response
+		}
+
+		pod.Spec.PreemptionPolicy = pc.PreemptionPolicy
+		pod.Spec.Priority = &pc.Value
+		pod.Spec.PriorityClassName = pc.Name
+
+		recorder.Eventf(tnt, corev1.EventTypeNormal, "TenantDefault", "Assigned Tenant default Priority Class %s to %s/%s", allowed.Default, pod.Namespace, pod.Name)
```

Comment:
Shouldn't we record this once we're sure the handler exited without any error?

At line 97 there's still a chance of getting an error.